executor/src/witgen/machines/machine_extractor.rs

```diff
@@ -64,18 +67,44 @@ pub fn split_out_machines<'a, T: FieldElement>(
     let mut base_identities = identities.clone();
     let mut extracted_prover_functions = HashSet::new();
     let mut id_counter = 0;
-    for id in &identities {
-        // Extract all witness columns in the RHS of the lookup.
-        let lookup_witnesses = match id {
-            Identity::Lookup(LookupIdentity { right, .. })
-            | Identity::PhantomLookup(PhantomLookupIdentity { right, .. })
-            | Identity::Permutation(PermutationIdentity { right, .. })
-            | Identity::PhantomPermutation(PhantomPermutationIdentity { right, .. }) => {
-                &refs_in_selected_expressions(right) & (&remaining_witnesses)
+
+    let all_connections = identities
+        .iter()
+        .filter_map(|i| Connection::try_from(*i).ok())
+        .collect::<Vec<_>>();
+
+    let mut fixed_lookup_connections = BTreeMap::new();
+
+    for connection in &all_connections {
+        // If the RHS only consists of fixed columns, record the connection and continue.
+        if connection.is_lookup()
+            && connection.right.selector.is_one()
+            && connection.right.expressions.iter().all(|e| {
+                try_to_simple_poly_ref(e)
+                    .map(|poly| poly.poly_id.ptype == PolynomialType::Constant)
+                    .unwrap_or(false)
+            })
+            && !connection.right.expressions.is_empty()
+        {
+            assert!(fixed_lookup_connections
+                .insert(connection.id, *connection)
+                .is_none());
+            if let Some(multiplicity) = connection.multiplicity_column {
+                remaining_witnesses.remove(&multiplicity);
```

Comment:
What if the multiplicity column occurs in a row-connected identity? I think we should at least check that.

Reply:
In this branch, we're talking about a fixed lookup, right? So the entire RHS is fixed columns. Do you mean that there could be constraints that connect those to witness columns?

If yes, I think that would be quite esoteric (those witness columns would not depend on an input and could be derived from the fixed columns at compile time), and it was also not checked before this PR (we `continue`d for any connection without witness columns on the RHS).